File: BaseLibrary/equip/MineGun.cs
using System;

namespace BaseLibrary.equip {
    public class MineGun : IClassEquipment{
        /// <inheritdoc />
        public int ID { get; private set; }

        /// <inheritdoc />
        public int COST { get; private set; }

        /// <summary>
        /// How many mines can be putted on map at the same time.
        /// </summary>
        public int MAX_MINES { get; private set; }

        /// <summary>
        /// Effect of mines. Their damaging and range of effect.
        /// </summary>
        public Zone[] ZONES { get; private set; }

        public MineGun(int ID, int COST, int MAX_MINES, Zone[] ZONES) {
            if (ZONES == null) throw new ArgumentNullException("zones");
            this.ID = ID;
            this.COST = COST;
            this.MAX_MINES = MAX_MINES;
            this.ZONES = new Zone[ZONES.Length];
            for (int i = 0; i < ZONES.Length; i++) {
                this.ZONES[i] = ZONES[i];
            }
        }

        /// <summary>
        /// Compare if both mine guns are same.
        /// </summary>
        /// <param name="other">Compared mine gun with this mine gun.</param>
        /// <returns></returns>
        protected bool Equals(MineGun other) {
            return ID == other.ID && COST == other.COST && MAX_MINES == other.MAX_MINES && Equals(ZONES, other.ZONES);
        }

        /// <inheritdoc />
        public override bool Equals(object obj) {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != GetType()) return false;
            return Equals((MineGun)obj);
        }

        /// <inheritdoc />
        public override int GetHashCode() {
            unchecked {
                int hashCode = ID;
                hashCode = (hashCode * 397) ^ COST;
                hashCode = (hashCode * 397) ^ MAX_MINES;
                hashCode = (hashCode * 397) ^ (ZONES?.GetHashCode() ?? 0);
                return hashCode;
            }
        }

        /// <inheritdoc />
        public override string ToString() {
            return $"{nameof(ID)}: {ID}, {nameof(COST)}: {COST}, {nameof(MAX_MINES)}: {MAX_MINES}, {nameof(ZONES)}: [{string.Join<Zone>(";", ZONES)}]";
        }
    }
}

File: Java-Client/RobotBattlefield-Client-Java/src/BaseLibrary/equipment/MineGun.java

package BaseLibrary.equipment;

import java.util.Arrays;

import BaseLibrary.communication.protocol.AProtocol;

public class MineGun implements IClassEquipment {
	private static final String NAME = "MINE_GUN";

	static {
		AProtocol.registerForDeserialize(NAME, MineGun.class);
	}

	@Override
	public String getCommandName() {
		return NAME;
	}

	public final int ID;

	public final int COST;

	/**
	 * How many mines can be put on map at the same time.
	 */
	public final int MAX_MINES;

	/**
	 * Effect of mines. Their damaging and range of effect.
	 */
	public final Zone[] ZONES;

	public MineGun(int ID, int COST, int MAX_MINES, Zone[] ZONES) {
		if (ZONES == null)
			throw new IllegalArgumentException("zones cannot be null.");
		this.ID = ID;
		this.COST = COST;
		this.MAX_MINES = MAX_MINES;
		this.ZONES = new Zone[ZONES.length];
		for (int i = 0; i < ZONES.length; i++) {
			this.ZONES[i] = ZONES[i];
		}
	}

	protected boolean equals(MineGun other) {
		if (COST != other.COST)
			return false;
		if (ID != other.ID)
			return false;
		if (MAX_MINES != other.MAX_MINES)
			return false;
		if (!Arrays.equals(ZONES, other.ZONES))
			return false;
		return true;
	}

	/**
	 * How many mines can be put on map at the same time.
	 */
	public int getMaxMines() {
		return MAX_MINES;
	}
	

	/**
	 * Effect of mines. Their damaging and range of effect.
	 */
	public Zone[] getZones() {
		return ZONES;
	}
	
	@Override
	public int getID() {
		return ID;
	}
	@Override
	public int getCost() {
		return COST;
	}
	
	@Override
	public String toString() {
		return "MineGun [ID=" + ID + ", COST=" + COST + ", MAX_MINES=" + MAX_MINES + ", ZONES=" + Arrays.toString(ZONES)
				+ "]";
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + COST;
		result = prime * result + ID;
		result = prime * result + MAX_MINES;
		result = prime * result + Arrays.hashCode(ZONES);
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		MineGun other = (MineGun) obj;
		return equals(other);
	}

}
